prov/efa/src/efa_fabric.c

```diff
@@ -479,13 +489,30 @@ static int efa_get_gdr_support(char *device_name)
 		goto out;
 	}
 
-	ret = fi_read_file(get_sysfs_path(), gdr_path, &gdr_support, sizeof(char));
-	if (ret < 0)
+	memset(gdr_support_str, 0, sizeof(gdr_support_str));
+	ret = fi_read_file(get_sysfs_path(), gdr_path,
+			   gdr_support_str, sizeof(gdr_support_str));
+	if (ret < 0) {
+		EFA_WARN(FI_LOG_FABRIC, "Cannot read sysfs file %s!\n", gdr_path);
+		goto out;
+	}
+
+	if (ret == 0) {
+		EFA_WARN(FI_LOG_FABRIC, "Sysfs file %s is empty\n", gdr_path);
+		ret = -FI_EINVAL;
 		goto out;
+	}
 
-	ret = atoi(&gdr_support);
+	ret = 0;
+	*gdr_support = strtol(gdr_support_str, NULL, 10);
```

Comment:
The point of using strtol instead of atoi was to use their error behavior.  See the `strtol()` manpage:

> If endptr is not NULL, strtol() stores the address of the first invalid
> character in *endptr.  If there were no digits at all, however, strtol()
> stores the original value of str in *endptr.  (Thus, if *str is not `\0'
> but **endptr is `\0' on return, the entire string was valid.)

Reply:
I see. Thanks! added a check of endptr in new revision.